File: apps/synqra-mvp/services/discovery_worker/unsubscribe_token.py
import base64
import hashlib
import hmac
import json
import os
from typing import Any


def create_unsubscribe_token(email: str, secret: str | None = None) -> str:
    normalized_email = _normalize_email(email)
    if not normalized_email:
        raise ValueError("email is required")

    payload_json = json.dumps({"email": normalized_email}, separators=(",", ":")).encode("utf-8")
    payload_b64 = _base64url_encode(payload_json)
    signature = _sign_payload(payload_b64, secret=secret)
    return f"{payload_b64}.{signature}"


def decode_unsubscribe_token(token: str, secret: str | None = None) -> str | None:
    if not isinstance(token, str) or "." not in token:
        return None

    payload_b64, signature = token.split(".", 1)
    expected_signature = _sign_payload(payload_b64, secret=secret)
    if not hmac.compare_digest(signature, expected_signature):
        return None

    try:
        payload_bytes = _base64url_decode(payload_b64)
        payload = json.loads(payload_bytes.decode("utf-8"))
    except Exception:
        return None

    email = payload.get("email")
    return _normalize_email(email)


def build_unsubscribe_link(email: str, base_url: str | None = None, secret: str | None = None) -> str:
    token = create_unsubscribe_token(email=email, secret=secret)
    root = (base_url or os.getenv("UNSUBSCRIBE_BASE_URL", "https://YOUR_DOMAIN")).rstrip("/")
    return f"{root}/ops/unsubscribe?t={token}"


def _normalize_email(value: Any) -> str | None:
    if not isinstance(value, str):
        return None
    normalized = value.strip().lower()
    return normalized if normalized else None


def _base64url_encode(value: bytes) -> str:
    return base64.urlsafe_b64encode(value).decode("ascii").rstrip("=")


def _base64url_decode(value: str) -> bytes:
    padding = "=" * ((4 - len(value) % 4) % 4)
    return base64.urlsafe_b64decode(value + padding)


def _sign_payload(payload_b64: str, secret: str | None = None) -> str:
    key = (secret or os.getenv("UNSUBSCRIBE_TOKEN_SECRET", "")).encode("utf-8")
    return hmac.new(key, payload_b64.encode("utf-8"), hashlib.sha256).hexdigest()

File: apps/synqra-mvp/lib/ops/unsubscribe-token.ts
import { createHmac, randomBytes, timingSafeEqual } from "node:crypto";

type TokenPayload = {
  email: string;
};

export function generateUnsubscribeToken(): string {
  return randomBytes(32).toString("hex");
}

export function validateUnsubscribeToken(token: string): boolean {
  try {
    return typeof token === "string" && /^[a-f0-9]{64}$/i.test(token.trim());
  } catch {
    return false;
  }
}

export function createUnsubscribeToken(email: string, secret = getSecret()): string {
  const normalized = normalizeEmail(email);
  if (!normalized) {
    throw new Error("Email is required");
  }

  const payloadB64 = base64UrlEncode(JSON.stringify({ email: normalized }));
  const signature = signPayload(payloadB64, secret);
  return `${payloadB64}.${signature}`;
}

export function decodeUnsubscribeToken(token: string, secret = getSecret()): string | null {
  const parts = token.split(".");
  if (parts.length !== 2) return null;

  const [payloadB64, signature] = parts;
  const expectedSignature = signPayload(payloadB64, secret);
  if (!timingSafeCompare(signature, expectedSignature)) {
    return null;
  }

  try {
    const parsed = JSON.parse(base64UrlDecode(payloadB64).toString("utf8")) as TokenPayload;
    return normalizeEmail(parsed.email);
  } catch {
    return null;
  }
}

export function buildUnsubscribeUrl(email: string, baseUrl = process.env.UNSUBSCRIBE_BASE_URL || "https://YOUR_DOMAIN"): string {
  const token = createUnsubscribeToken(email);
  return `${baseUrl.replace(/\/+$/, "")}/ops/unsubscribe?t=${encodeURIComponent(token)}`;
}

function normalizeEmail(value: string | null | undefined): string | null {
  if (typeof value !== "string") return null;
  const normalized = value.trim().toLowerCase();
  return normalized.length > 0 ? normalized : null;
}

function signPayload(payloadB64: string, secret: string): string {
  return createHmac("sha256", secret).update(payloadB64).digest("hex");
}

function getSecret(): string {
  return process.env.UNSUBSCRIBE_TOKEN_SECRET || "";
}

function base64UrlEncode(value: string): string {
  return Buffer.from(value, "utf8").toString("base64url");
}

function base64UrlDecode(value: string): Buffer {
  return Buffer.from(value, "base64url");
}

function timingSafeCompare(left: string, right: string): boolean {
  const leftBuffer = Buffer.from(left);
  const rightBuffer = Buffer.from(right);
  if (leftBuffer.length !== rightBuffer.length) return false;
  return timingSafeEqual(leftBuffer, rightBuffer);
}
